artemis/src/main/java/tech/pegasys/artemis/DepositCommand.java

```diff
@@ -81,44 +99,144 @@ public void generate(
+      try {
+        futures =
+            validatorKeyPairs.stream()
+                .map(
+                    keyPair -> {
+                      final String yamlFormattedString = getYamlFormattedString(keyPair);
+                      if (encryptKeys) {
+                        generateKeystores(
+                            keyPair, keyStoreOutputDirectory, counterSuffix.incrementAndGet());
+                      } else if (printOnStandardOut) {
+                        System.out.println(yamlFormattedString);
+                      } else {
+                        yamlStringBuilder
+                            .append(yamlFormattedString)
+                            .append(System.lineSeparator());
+                      }
+
+                      return sendDeposit(
+                          sender,
+                          keyPair.getLeft(),
+                          keyPair.getRight().getPublicKey(),
+                          params.amount);
+                    })
+                .collect(Collectors.toList());
 
-      try (keyWriter) {
-        for (int i = 0; i < validatorCount; i++) {
-          final BLSKeyPair validatorKey = BLSKeyPair.random();
-          final BLSKeyPair withdrawalKey = BLSKeyPair.random();
-
-          keyWriter.println(
-              String.format(
-                  "- {privkey: '%s', pubkey: '%s', withdrawalPrivkey: '%s', withdrawalPubkey: '%s'}",
-                  validatorKey.getSecretKey().getSecretKey().toBytes(),
-                  validatorKey.getPublicKey().toBytesCompressed(),
-                  withdrawalKey.getSecretKey().getSecretKey().toBytes(),
-                  withdrawalKey.getPublicKey().toBytesCompressed()));
-          futures.add(
-              sendDeposit(sender, validatorKey, withdrawalKey.getPublicKey(), params.amount));
+      } finally {
+        if (!encryptKeys && !printOnStandardOut) {
+          // write keys yaml to file
+          Files.writeString(Path.of(outputFile), yamlStringBuilder.toString());
         }
       }
-      SafeFuture.allOf(futures.toArray(SafeFuture[]::new)).get(2, TimeUnit.MINUTES);
+
+      waitForTransactionReceipts(futures);
     } catch (final Throwable t) {
       STDOUT.log(
           Level.FATAL,
           "Failed to send deposit transaction: " + t.getClass() + ": " + t.getMessage());
-      System.exit(1); // Web3J creates a non-daemon thread we can't shut down. :(
+      exit(1);
     }
-    System.exit(0); // Web3J creates a non-daemon thread we can't shut down. :(
+    exit(0);
+  }
+
+  @VisibleForTesting
+  void waitForTransactionReceipts(final List<SafeFuture<TransactionReceipt>> futures)
+      throws InterruptedException, ExecutionException, TimeoutException {
+    SafeFuture.allOf(futures.toArray(SafeFuture[]::new)).get(2, TimeUnit.MINUTES);
+  }
+
+  @VisibleForTesting
+  void exit(final int status) {
+    System.exit(status); // Web3J creates a non-daemon thread we can't shut down. :(
+  }
```

Comment:
Rather than overriding this method in tests, it would be better to pass in a `Consumer<Integer> shutdownFunction` to the constructor (ie prefer delegation over inheritance).  You can pass a mock in tests and `System::exit` in production.

You will likely need two constructors for `DepositCommand` - a noarg one that passes `System::exit` and one that accepts the consumer that tests can use.

Reply:
Implemented functional consumer interface.